kcoin/genesis/genesis_generator.go

```diff
@@ -0,0 +1,404 @@
+	}
+	genesis.ExtraData = make([]byte, 32)
+	if len(extra) > 32 {
+		extra = extra[:32]
+	}
+	genesis.ExtraData = append([]byte(extra), genesis.ExtraData[len(extra):]...)
+}
+
+func setConsensusEngine(consensusEngine string, genesis *core.Genesis) {
+	switch consensusEngine {
+	case TendermintConsensus:
+		genesis.Config.Tendermint = &params.TendermintConfig{Rewarded: true}
+	}
+}
+
+func setNetwork(network string, genesis *core.Genesis) {
+	switch network {
+	case MainNetwork:
+		genesis.Config.ChainID = params.MainnetChainConfig.ChainID
+	case TestNetwork:
+		genesis.Config.ChainID = params.TestnetChainConfig.ChainID
+	case OtherNetwork:
+		genesis.Config.ChainID = new(big.Int).SetUint64(uint64(rand.Intn(65536)))
+	}
+}
+
+func validateOptions(options GenesisOptions) (*validGenesisOptions, error) {
+	network, err := createNetwork(options.Network)
+	if err != nil {
+		return nil, err
+	}
+
+	maxNumValidators, err := createMaxNumValidators(options.MaxNumValidators)
+	if err != nil {
+		return nil, err
+	}
+
+	unbondingPeriod, err := createUnbondingPeriod(options.UnbondingPeriod)
+	if err != nil {
+		return nil, err
+	}
+
+	walletAddressValidator, err := createWalletAddress(options.WalletAddressGenesisValidator)
+	if err != nil {
+		return nil, err
+	}
+
+	validPrefundedAccounts, err := createPrefundedAccounts(options.PrefundedAccounts)
+	if err != nil {
+		return nil, err
+	}
+
+	if !prefundedIncludesValidatorWallet(validPrefundedAccounts, walletAddressValidator) {
+		return nil, ErrWalletAddressValidatorNotInPrefundedAccounts
+	}
+
+	consensusEngine := TendermintConsensus
+	if options.ConsensusEngine != "" {
+		consensusEngine, err = createConsensusEngine(options.ConsensusEngine)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	owner := &DefaultSmartContractsOwner
+	if options.SmartContractsOwner != "" {
+		strAddr := options.SmartContractsOwner
+
+		owner, err = createWalletAddress(strAddr)
+		if err != nil {
+			return nil, ErrInvalidContractsOwnerAddress
+		}
+	}
+
+	return &validGenesisOptions{
+		network:                       network,
+		maxNumValidators:              maxNumValidators,
+		unbondingPeriod:               unbondingPeriod,
+		walletAddressGenesisValidator: walletAddressValidator,
+		prefundedAccounts:             validPrefundedAccounts,
+		consensusEngine:               consensusEngine,
+		smartContractsOwner:           owner,
+	}, nil
+}
+
+func addBatchOfPrefundedAccounts(genesis *core.Genesis) {
+	// Add a batch of precompile balances to avoid them getting deleted
+	for i := int64(0); i < 256; i++ {
+		genesis.Alloc[common.BigToAddress(big.NewInt(i))] = core.GenesisAccount{Balance: big.NewInt(1)}
+	}
+}
+
+func setPrefundedAccounts(validPrefundedAccounts []*validPrefundedAccount, genesis *core.Genesis) {
+	for _, vAccount := range validPrefundedAccounts {
+		genesis.Alloc[*vAccount.walletAddress] = core.GenesisAccount{
+			Balance: vAccount.balance,
+		}
+	}
+}
+
+func createNetwork(network string) (string, error) {
```

Comment:
`get`, `map` or similar on these func names, I think `create` implies creating a object or a resource, this just maps between 2 values

Reply:
Thanks, I had some issues with the naming on this functions.